Add sanitizer support on FreeBSD

valpackett opened the pull request:
Restarting #47337. Everything is better now, no more weird llvm problems, well not everything:

Unfortunately, the sanitizers don't have proper support for versioned symbols (https://github.com/google/sanitizers/issues/628), so `libc`'s usage of `stat@FBSD_1.0` and so on explodes, e.g. in calling `std::fs::metadata`.

Building std (now easy thanks to cargo `-Zbuild-std`) and libc with `freebsd12/13` config via the `LIBC_CI=1` env variable is a good workaround…

```
LIBC_CI=1 RUSTFLAGS="-Z sanitizer=address" cargo +san-test -Zbuild-std run --target x86_64-unknown-freebsd --verbose
```

…*except* std won't build because there's no `st_lspare` in the ino64 version of the struct, so an std patch is required:

```diff
--- i/src/libstd/os/freebsd/fs.rs
+++ w/src/libstd/os/freebsd/fs.rs
@@ -66,8 +66,6 @@ pub trait MetadataExt {
     fn st_flags(&self) -> u32;
     #[stable(feature = "metadata_ext2", since = "1.8.0")]
     fn st_gen(&self) -> u32;
-    #[stable(feature = "metadata_ext2", since = "1.8.0")]
-    fn st_lspare(&self) -> u32;
 }
 
 #[stable(feature = "metadata_ext", since = "1.1.0")]
@@ -136,7 +134,4 @@ impl MetadataExt for Metadata {
     fn st_flags(&self) -> u32 {
         self.as_inner().as_inner().st_flags as u32
     }
-    fn st_lspare(&self) -> u32 {
-        self.as_inner().as_inner().st_lspare as u32
-    }
 }
```

I guess std could like.. detect that `libc` isn't built for the old ABI, and replace the implementation of `st_lspare` with a panic?

valpackett committed:
Add sanitizer support on FreeBSD

rust-highfive commented:
r? @oli-obk

(rust_highfive has picked a reviewer for you, use r? to override)

oli-obk commented:
That std patch by itself shouldn't be necessary, https://github.com/rust-lang/rust/blob/d8bdb3fdcbd88eb16e1a6669236122c41ed2aed3/src/libstd/os/freebsd/raw.rs#L78 is not conditional, so if that is missing on the C-header version of the same struct, it needs to be compiled out here

myfreeweb commented:
> it needs to be compiled out here

That's what I'm wondering how to do..

> I guess std could like.. detect that libc isn't built for the old ABI, and replace the implementation of st_lspare with a panic?

(just excluding the method would be fine too, almost nobody really seems to use that ext thing anyway)

I guess the only way is to repeat what libc's build.rs does with `LIBC_CI` / `freebsd-version`..?

myfreeweb commented:
> is not conditional

Oh — the thing is, the accessor does not use the `src/libstd/os/freebsd/raw.rs` struct, it goes into the `libc` struct!

Maybe that should be changed, and then if we detect the `libc` being built for the newer ABI, either do not compile the `ext` at all (it won't match) or make another version that matches.

oli-obk commented:
If the libc crate can figure this out, maybe we can, too? Or is this done in a build script?

myfreeweb commented:
Yes: https://github.com/rust-lang/libc/blob/dce0ec0adb8f8ef711ea6b54e90429972f6077a0/build.rs#L21-L34

oli-obk commented:
cc @alexcrichton 

while implementation wise we can definitely add more things to the libstd build script, I don't know anything about this area and how to proceed best. `libc` has some `libstd` special compile-time logic already, so I think we need someone who already knows a bit about this link between the two libs.

alexcrichton commented:
Unfortunately I'm not sure I'll be able to help much here. It sounds like this is tied up in the question of what FreeBSD version is targeted for targets like `x86_64-unknown-freebsd`, and that's one which I don't think has a great answer in libstd as a precompiled binary or as source. It sounds like what needs to happen is *somehow* libstd itself needs to have a `#[cfg]` to switch between implementations but that cfg doesn't exist today easily. While libstd could replicate what libc is doing I'm not sure if that's the best answer.

Overally I don't have a great answer here, sorry :(. I think it would work to duplicate libc's build script into libstd, but that's not exactly a great solution unfortunately.

myfreeweb commented:
The great solution would be to have target versions (`--target x86_64-unknown-freebsd13`), but for now I guess we can only do something like that. Maybe with a less bad variable name than `LIBC_CI`..

oli-obk commented:
> The great solution would be to have target versions (`--target x86_64-unknown-freebsd13`), but for now I guess we can only do something like that. Maybe with a less bad variable name than `LIBC_CI`..

Hmm... I couldn't find any preexisting versioning like this in https://github.com/rust-lang/rust/tree/e59b08e62ea691916d2f063cac5aab4634128022/src/librustc_target/spec so yea, that would be a major undertaking. How about going with something really obvious like `FREEBSD_12_13_SANITIZER_SUPPORT_HACK`?

valpackett committed:
Add RUST_STD_FREEBSD_12_ABI env variable

Unfortunately, sanitizers do not support versioned symbols[1],
so they break filesystem access via the legacy, pre-ino64 ABI.

To use sanitizers on FreeBSD >= 12, we need to build the libc
crate with LIBC_CI=1 to use the new ABI -- including the libc
used for std. But that removes the st_lspare field std was
expecting for the deprecated metadata extension.

Add a way to skip that field to allow the build to work.

[1]: https://github.com/google/sanitizers/issues/628

bors commented:
:umbrella: The latest upstream changes (presumably #74844) made this pull request unmergeable. Please resolve the merge conflicts.

oli-obk commented on library/std/src/os/freebsd/fs.rs:
```diff
@@ -74,6 +75,10 @@ pub trait MetadataExt {
+        // This method would just return nonsense.
+        #[cfg(freebsd12)]
+        panic!("as_raw_stat not supported with FreeBSD 12 ABI");
         unsafe { &*(self.as_inner().as_inner() as *const libc::stat as *const raw::stat) }
```
you also need to make this code `#[cfg(not(freebsd12))]` otherwise you'll get warnings or even errors when compiling libstd

oli-obk commented on library/std/src/os/freebsd/fs.rs:
```diff
@@ -136,6 +141,7 @@ impl MetadataExt for Metadata {
         self.as_inner().as_inner().st_flags as u32
     }
+    #[cfg(not(freebsd12))]
     fn st_lspare(&self) -> u32 {
```
huh, this function outright missing doesn't cause errors elsewhere in if the flag is active?

myfreeweb replied:
Definitely not in Rust itself. Maaaaaaybe there's some user code out there using it, but I wouldn't be surprised if nobody ever touched this obscure field on this deprecated way to access the raw stat data.. Okay I guess it would be better to panic instead

oli-obk commented on library/std/build.rs:
```diff
@@ -16,6 +16,9 @@ fn main() {
         println!("cargo:rustc-link-lib=pthread");
+        if env::var("RUST_STD_FREEBSD_12_ABI").is_ok() {
+            println!("cargo:rustc-cfg=freebsd12");
+        } else { unreachable!()}
```
that's not quite true, but also not terribly important. Just panic with a helpful message?

myfreeweb replied:
oops, I was testing, this definitely should not be there

oli-obk commented:
sorry, I didn't see your force push

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit baaf0843c9743a7f36ebbff05cddee3a838d5359 has been approved by `oli-obk`

<!-- @bors r=oli-obk baaf0843c9743a7f36ebbff05cddee3a838d5359 -->
<!-- homu: {"type":"Approved","sha":"baaf0843c9743a7f36ebbff05cddee3a838d5359","approver":"oli-obk"} -->

Manishearth commented:
@bors r-

https://github.com/rust-lang/rust/pull/74987#issuecomment-667338084

tmiasko commented on library/std/src/os/freebsd/fs.rs:
```diff
@@ -137,6 +144,10 @@ impl MetadataExt for Metadata {
         self.as_inner().as_inner().st_flags as u32
     }
     fn st_lspare(&self) -> u32 {
+        #[cfg(freebsd12)]
+        panic!("st_lspare not supported with FreeBSD 12 ABI");
+        #[cfg(not(freebsd12))]
+        #[cfg(not(freebsd12))]
         self.as_inner().as_inner().st_lspare as u32
     }
```
Unfortunately, in this case the rustfmt fix suggestion is bogus. Following should work:

```suggestion
    #[cfg(freebsd12)]
    fn st_lspare(&self) -> u32 {
        panic!("st_lspare not supported with FreeBSD 12 ABI");
    }
    #[cfg(not(freebsd12))]
    fn st_lspare(&self) -> u32 {
        self.as_inner().as_inner().st_lspare as u32
    }
```

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit ddbc45673b11f706e94fa17730eaa5c113f224d7 has been approved by `oli-obk`

<!-- @bors r=oli-obk ddbc45673b11f706e94fa17730eaa5c113f224d7 -->
<!-- homu: {"type":"Approved","sha":"ddbc45673b11f706e94fa17730eaa5c113f224d7","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit ddbc45673b11f706e94fa17730eaa5c113f224d7 with merge e9d65c53543dee07c4dd26407716f6083c43c4c1...
<!-- homu: {"type":"BuildStarted","head_sha":"ddbc45673b11f706e94fa17730eaa5c113f224d7","merge_sha":"e9d65c53543dee07c4dd26407716f6083c43c4c1"} -->

rust-log-analyzer commented:
The job `dist-various-1` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/965868303) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/e9d65c53543dee07c4dd26407716f6083c43c4c1/checks/965868303/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/e9d65c53543dee07c4dd26407716f6083c43c4c1/checks/965868303/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/965884796)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/965884796","builder_name":"checks-actions"} -->

tmiasko commented:
Spurious network error:
```
warning: spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/adler/0.2.3/download`, got 502
error: failed to download from `https://crates.io/api/v1/crates/miniz_oxide/0.4.0/download`
```

oli-obk commented:
@bors retry

bors commented:
:hourglass: Testing commit ddbc45673b11f706e94fa17730eaa5c113f224d7 with merge c18def13276b39d8270d559735e824f4167c307b...
<!-- homu: {"type":"BuildStarted","head_sha":"ddbc45673b11f706e94fa17730eaa5c113f224d7","merge_sha":"c18def13276b39d8270d559735e824f4167c307b"} -->

rust-log-analyzer commented:
The job `dist-x86_64-freebsd` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/967248256) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/c18def13276b39d8270d559735e824f4167c307b/checks/967248256/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/c18def13276b39d8270d559735e824f4167c307b/checks/967248256/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
In file included from /checkout/src/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_freebsd.cpp:50:
/usr/local/x86_64-unknown-freebsd11/usr/include/sys/timeb.h:42:2: warning: "this file includes <sys/timeb.h> which is deprecated" [-W#warnings]
#warning "this file includes <sys/timeb.h> which is deprecated"
[ 73%] Building CXX object lib/sanitizer_common/CMakeFiles/RTSanitizerCommon.x86_64.dir/sanitizer_procmaps_common.cpp.o
1 warning generated.
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
---
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
[ 95%] Building ASM object lib/asan/CMakeFiles/RTAsan.x86_64.dir/asan_interceptors_vfork.S.o
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
/checkout/src/llvm-project/compiler-rt/lib/asan/asan_interceptors_vfork.S:12:1: warning: DWARF2 only supports one section per compilation unit
.section .note.GNU-stack,"",%progbits
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
[ 95%] Built target RTSanitizerCommon.x86_64
---
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_fd.h:36:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_fd.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:134: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_fd.cpp.o' failed
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_clock.cpp:13:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_clock.cpp.o] Error 1
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:62: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_clock.cpp.o' failed
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface_java.cpp:14:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_mutexset.cpp:13:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_external.cpp:12:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:110: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_external.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:302: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_java.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:422: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mutexset.cpp.o' failed
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
#include <sys/_types.h>
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
1 error generated.
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_external.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_java.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mutexset.cpp.o] Error 1
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_debugging.cpp:14:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_report.h:15:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_symbolizer.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_vector.h:18:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_vector.h:18:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator_internal.h:16:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_debugging.cpp.o] Error 1
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:86: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_debugging.cpp.o' failed
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_flags.cpp:17:
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
---
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator_internal.h:16:
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
         ^~~~~~~~~~~~~~
1 error generated.
1 error generated.
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.cpp:15:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_rtl.h:28:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
         ^~~~~~~~~~~~~~
1 error generated.
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_allocator.h:23:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
#include <sys/_types.h>
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_procmaps.h:23:
         ^~~~~~~~~~~~~~
1 error generated.
1 error generated.
1 error generated.
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
In file included from /checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_linux.h:20:
#include <sys/_types.h>
1 error generated.
1 error generated.
/checkout/src/llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_platform_limits_freebsd.h:30:10: fatal error: 'sys/_types.h' file not found
#include <sys/_types.h>
1 error generated.
1 error generated.
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_flags.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_ann.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mman.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_atomic.cpp.o] Error 1
make[3]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mutex.cpp.o] Error 1
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:158: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_flags.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:254: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_ann.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:230: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:374: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mman.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:278: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_interface_atomic.cpp.o' failed
lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/build.make:398: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rtl/tsan_mutex.cpp.o' failed
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
make[2]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/all] Error 2
CMakeFiles/Makefile2:4771: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/all' failed
make[1]: *** [lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rule] Error 2
CMakeFiles/Makefile2:4783: recipe for target 'lib/tsan/CMakeFiles/clang_rt.tsan-x86_64.dir/rule' failed
make: *** [clang_rt.tsan-x86_64] Error 2
Makefile:1711: recipe for target 'clang_rt.tsan-x86_64' failed
command did not execute successfully, got: exit code: 2


build script failed, must exit now', /cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.42/src/lib.rs:861:5
 finished in 25.725
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap dist --host x86_64-unknown-freebsd --target x86_64-unknown-freebsd
Build completed unsuccessfully in 0:07:40
== clock drift check ==
== clock drift check ==
  local time: Mon Aug 10 15:18:04 UTC 2020
  network time: Mon, 10 Aug 2020 15:18:04 GMT
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (3425) (node)
Terminate orphan process: pid (3453) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/967303424)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/967303424","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job `dist-various-1` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/965868303) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/e9d65c53543dee07c4dd26407716f6083c43c4c1/checks/965868303/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/e9d65c53543dee07c4dd26407716f6083c43c4c1/checks/965868303/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

tmiasko commented:
Can we retry this? The #75376 should address tsan build failure.

oli-obk commented:
@bors retry

bors commented:
:hourglass: Testing commit ddbc45673b11f706e94fa17730eaa5c113f224d7 with merge 80fb3f3139c7dee7f211964c6a0b3ccb04b83d5e...
<!-- homu: {"type":"BuildStarted","head_sha":"ddbc45673b11f706e94fa17730eaa5c113f224d7","merge_sha":"80fb3f3139c7dee7f211964c6a0b3ccb04b83d5e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/988237496), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36005)
Approved by: oli-obk
Pushing 80fb3f3139c7dee7f211964c6a0b3ccb04b83d5e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36005","checks-actions":"https://github.com/rust-lang-ci/rust/runs/988237496"},"merge_sha":"80fb3f3139c7dee7f211964c6a0b3ccb04b83d5e"} -->
